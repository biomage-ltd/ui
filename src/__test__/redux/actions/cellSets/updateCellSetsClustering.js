import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';
import updateCellSetsClustering from '../../../../redux/actions/cellSets/updateCellSetsClustering';
import initialState from '../../../../redux/reducers/cellSets/initialState';
import sendWork from '../../../../utils/sendWork';

enableFetchMocks();
const mockStore = configureStore([thunk]);
jest.mock('../../../../utils/sendWork', () => ({
  __esModule: true, // this property makes it work
  default: jest.fn(),
}));

describe('updateCellSetsClustering action', () => {
  const experimentId = '1234';

  beforeEach(() => {
    const response = new Response(JSON.stringify({}));

    fetchMock.resetMocks();
    fetchMock.doMock();
    fetchMock.mockResolvedValueOnce(response);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('Does not dispatch on loading state', async () => {
    const store = mockStore({ cellSets: { loading: true, error: false } });
    store.dispatch(updateCellSetsClustering(experimentId));
    expect(store.getActions().length).toEqual(0);
  });

  it('Does not dispatch on error state', async () => {
    const store = mockStore({ cellSets: { loading: false, error: true } });
    store.dispatch(updateCellSetsClustering(experimentId));
    expect(store.getActions().length).toEqual(0);
  });

  it('Dispatches all required actions to update cell sets clustering.', async (done) => {
    const store = mockStore({ cellSets: { ...initialState, loading: false } });
    sendWork.mockImplementation(() => {
      const resolveWith = {
        results:
          [
            {
              body: JSON.stringify({
                name: 'one', color: '#ff0000', cellIds: ['1', '2', '3'],
              }),
            },
          ],
      };

      return new Promise((resolve) => resolve(resolveWith));
    });

    const flushPromises = () => new Promise(setImmediate);

    store.dispatch(updateCellSetsClustering(experimentId, 0.5));

    expect(sendWork).toHaveBeenCalledTimes(1);
    expect(sendWork).toHaveBeenCalledWith(experimentId, 30, {
      name: 'ClusterCells',
      cellSetName: 'Louvain clusters',
      type: 'louvain',
      cellSetKey: 'louvain',
      config: { resolution: 0.5 },
    });

    await flushPromises();

    const loadingAction = store.getActions()[0];
    expect(loadingAction).toMatchSnapshot();

    const loadedAction = store.getActions()[1];
    expect(loadedAction).toMatchSnapshot();

    const savedAction = store.getActions()[2];
    expect(savedAction).toMatchSnapshot();
    done();
  });

  it('Dispatches error action when the reset fails', async () => {
    const store = mockStore({ cellSets: { ...initialState, loading: false } });
    sendWork.mockImplementation(() => {
      const resolveWith = {
        results: { error: 'The backend returned an error' },
      };

      return new Promise((resolve) => resolve(resolveWith));
    });

    const flushPromises = () => new Promise(setImmediate);

    store.dispatch(updateCellSetsClustering(experimentId, 0.5));

    expect(sendWork).toHaveBeenCalledTimes(1);
    expect(sendWork).toHaveBeenCalledWith(experimentId, 30, {
      name: 'ClusterCells',
      cellSetName: 'Louvain clusters',
      type: 'louvain',
      cellSetKey: 'louvain',
      config: { resolution: 0.5 },
    });

    await flushPromises();

    const loadingAction = store.getActions()[0];
    expect(loadingAction).toMatchSnapshot();

    const errorAction = store.getActions()[1];
    expect(errorAction).toMatchSnapshot();
  });
});
