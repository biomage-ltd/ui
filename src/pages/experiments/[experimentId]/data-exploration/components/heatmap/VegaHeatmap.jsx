import React from 'react';
import PropTypes from 'prop-types';
import { Vega } from 'react-vega';
import { Element } from 'react-scroll';
import './Heatmap.module.css';

const VegaHeatmap = (props) => {
  const {
    spec, showAxes, rowsNumber, signalListeners, data, width, height,
  } = props;

  const axes = [
    {
      domain: false,
      orient: 'left',
      scale: 'y',
    },
  ];

  const getAdjustedHeight = () => {
    const minHeightPerGene = 6;
    const heightPerGene = 30 - rowsNumber / 2;
    if (heightPerGene < minHeightPerGene || heightPerGene * rowsNumber > height) {
      return height;
    }
    return heightPerGene * rowsNumber;
  };

  const getAxes = () => {
    if (showAxes) {
      return axes;
    }
    return [];
  };

  const vegaSpec = {
    ...spec,
    width: width * 0.98,
    height: getAdjustedHeight() * 0.9,
    autosize: { type: 'fit' },
    axes: [...spec.axes, ...getAxes()],
    data: spec.data.map((datum) => ({
      ...datum,
      values: data[datum.name],
    })),
  };

  return (
    <Element
      className='element'
      id='heatmap-container'
    >
      <Vega
        spec={vegaSpec}
        signalListeners={signalListeners}
        actions={false}
        onNewView={(e) => console.log(e)}
      />
    </Element>
  );
};

VegaHeatmap.defaultProps = {
  showAxes: true,
};

VegaHeatmap.propTypes = {
  spec: PropTypes.object.isRequired,
  showAxes: PropTypes.bool,
  rowsNumber: PropTypes.number.isRequired,
  data: PropTypes.object.isRequired,
  defaultWidth: PropTypes.number.isRequired,
  signalListeners: PropTypes.object.isRequired,
  width: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
};

export default VegaHeatmap;
