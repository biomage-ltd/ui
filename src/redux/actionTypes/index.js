const GENE_LIST = 'GENE_LIST';
const LOAD_GENE_LIST = `${GENE_LIST}.LOAD`;
const UPDATE_GENE_LIST = `${GENE_LIST}.UPDATE`;

const SELECTED_GENES = 'SELECTED_GENES';

const BUILD_HEATMAP_SPEC = 'BUILD_HEATMAP_SPEC';
const UPDATE_GENE_EXPRESSION = 'UPDATE_GENE_EXPRESSION';
const UPDATE_HEATMAP_SPEC = 'UPDATE_HEATMAP_SPEC';

const SET_FOCUSED_GENE = 'SET_FOCUSED_GENE';

export {
  LOAD_GENE_LIST, UPDATE_GENE_LIST, SELECTED_GENES, BUILD_HEATMAP_SPEC,
  UPDATE_GENE_EXPRESSION, UPDATE_HEATMAP_SPEC,
  SET_FOCUSED_GENE,
};
