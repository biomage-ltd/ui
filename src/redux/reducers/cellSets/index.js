import {
  CELL_SETS_LOADING, CELL_SETS_LOADED,
  CELL_SETS_CREATE,
  CELL_SETS_UPDATE_PROPERTY, CELL_SETS_UPDATE_HIERARCHY, CELL_SETS_SET_SELECTED,
  CELL_SETS_DELETE,
  CELL_SETS_HIDE, CELL_SETS_UNHIDE, CELL_SETS_UNHIDE_ALL,
  CELL_SETS_ERROR,
} from '../../actionTypes/cellSets';

import initialState from './initialState';

import cellSetsLoading from './cellSetsLoading';
import cellSetsLoaded from './cellSetsLoaded';
import cellSetsUpdateProperty from './cellSetsUpdateProperty';
import cellSetsDelete from './cellSetsDelete';
import cellSetsUpdateHierarchy from './cellSetsUpdateHierarchy';
import cellSetsCreate from './cellSetsCreate';
import cellSetsSetSelected from './cellSetsSetSelected';
import cellSetsError from './cellSetsError';
import { cellSetsHide, cellSetsUnhide, cellSetsUnhideAll } from './cellSetsHideUnhide';

const cellSetsReducer = (state = initialState, action) => {
  switch (action.type) {
    case CELL_SETS_LOADING: {
      return cellSetsLoading(state, action);
    }

    case CELL_SETS_LOADED: {
      return cellSetsLoaded(state, action);
    }

    case CELL_SETS_UPDATE_PROPERTY: {
      return cellSetsUpdateProperty(state, action);
    }

    case CELL_SETS_DELETE: {
      return cellSetsDelete(state, action);
    }

    case CELL_SETS_UPDATE_HIERARCHY: {
      return cellSetsUpdateHierarchy(state, action);
    }

    case CELL_SETS_CREATE: {
      return cellSetsCreate(state, action);
    }

    case CELL_SETS_SET_SELECTED: {
      return cellSetsSetSelected(state, action);
    }

    case CELL_SETS_HIDE: {
      return cellSetsHide(state, action);
    }

    case CELL_SETS_UNHIDE: {
      return cellSetsUnhide(state, action);
    }

    case CELL_SETS_UNHIDE_ALL: {
      return cellSetsUnhideAll(state, action);
    }

    case CELL_SETS_ERROR: {
      return cellSetsError(state, action);
    }

    default: {
      return state;
    }
  }
};

export default cellSetsReducer;
