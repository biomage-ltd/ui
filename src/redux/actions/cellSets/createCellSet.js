import { v4 as uuidv4 } from 'uuid';
import messages from '../../../components/notification/messages';
import { CELL_SETS_CREATE } from '../../actionTypes/cellSets';
import saveCellSets from './saveCellSets';
import pushNotificationMessage from '../../../utils/pushNotificationMessage';

const createCellSet = (experimentId, name, color, cellIds) => async (dispatch, getState) => {
  const {
    loading, error,
  } = getState().cellSets;

  if (loading || error) {
    return null;
  }

  const data = {
    key: uuidv4(),
    name,
    color,
    cellIds: new Set([...cellIds].map((id) => parseInt(id, 10))),
  };

  if (data.cellIds.size === 0) {
    pushNotificationMessage('info', messages.emptyClusterNotCreated, 5);
    return;
  }

  dispatch({
    type: CELL_SETS_CREATE,
    payload: {
      experimentId,
      ...data,
    },
  });

  await dispatch(saveCellSets(experimentId));
  pushNotificationMessage('info', messages.newClusterCreated, 5);
};

export default createCellSet;
