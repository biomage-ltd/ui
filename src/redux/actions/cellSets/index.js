import createCellSet from './createCellSet';
import deleteCellSet from './deleteCellSet';
import resetCellSets from './resetCellSets';

import loadCellSets from './loadCellSets';
import saveCellSets from './saveCellSets';
import updateCellSetHierarchy from './updateCellSetHierarchy';
import updateCellSetProperty from './updateCellSetProperty';
import updateCellSetSelected from './updateCellSetSelected';
import setCellSetHiddenStatus from './setCellSetHiddenStatus';
import unhideAllCellSets from './unhideAllCellSets';

export {
  createCellSet,
  deleteCellSet,
  resetCellSets,
  loadCellSets,
  saveCellSets,
  updateCellSetHierarchy,
  updateCellSetSelected,
  updateCellSetProperty,
  setCellSetHiddenStatus,
  unhideAllCellSets,
};
