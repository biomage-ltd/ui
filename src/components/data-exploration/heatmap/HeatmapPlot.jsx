import React, {
  useRef, useEffect, useState, useCallback,
} from 'react';
import PropTypes from 'prop-types';
import { useSelector, useDispatch } from 'react-redux';
import {
  Empty, Typography, Skeleton,
} from 'antd';
import _ from 'lodash';
import spec from '../../../utils/heatmapSpec';
import VegaHeatmap from './VegaHeatmap';
import PlatformError from '../../PlatformError';
import { updateCellInfo } from '../../../redux/actions/cellInfo';
import { loadGeneExpression } from '../../../redux/actions/genes';
import { loadCellSets } from '../../../redux/actions/cellSets';
import { loadComponentConfig } from '../../../redux/actions/componentConfig';
import populateHeatmapData from '../../plots/helpers/populateHeatmapData';
import Loader from '../../Loader';

const COMPONENT_TYPE = 'interactiveHeatmap';
const { Text } = Typography;

const HeatmapPlot = (props) => {
  const {
    experimentId, width, height,
  } = props;

  const dispatch = useDispatch();

  const loadingGenes = useSelector((state) => state.genes.expression.loading);
  const selectedGenes = useSelector((state) => state.genes.expression.views[COMPONENT_TYPE]?.data);
  const [vegaData, setVegaData] = useState(null);
  const [vegaSpec, setVegaSpec] = useState(spec);

  const expressionData = useSelector((state) => state.genes.expression);
  const hoverCoordinates = useRef({});

  const cellSets = useSelector((state) => state.cellSets);
  const {
    hierarchy, properties, hidden, loading: cellSetsLoading,
  } = cellSets;

  const heatmapSettings = useSelector(
    (state) => state.componentConfig[COMPONENT_TYPE]?.config,
  ) || {};

  const {
    selectedTracks, groupedTracks, expressionValue, legendIsVisible,
  } = heatmapSettings;

  const { error } = expressionData;
  const viewError = useSelector((state) => state.genes.expression.views[COMPONENT_TYPE]?.error);

  const [maxCells, setMaxCells] = useState(1000);

  const setDataDebounce = useCallback(_.debounce((data) => {
    setVegaData(data);
  }, 1500, { leading: true }), []);

  /**
   * Loads cell set on initial render if it does not already exist in the store.
   */
  useEffect(() => {
    dispatch(loadCellSets(experimentId));
  }, []);

  useEffect(() => {
    if (!_.isEmpty(heatmapSettings)) {
      return;
    }

    dispatch(loadComponentConfig(experimentId, COMPONENT_TYPE, COMPONENT_TYPE));
  }, [heatmapSettings]);

  useEffect(() => {
    if (hierarchy.length === 0 || cellSetsLoading) {
      return;
    }

    const legends = legendIsVisible ? spec.legends : [];
    setVegaSpec({ ...spec, legends });
  }, [legendIsVisible]);

  useEffect(() => {
    if (!selectedGenes || selectedGenes.length === 0) {
      return;
    }

    if (_.intersection(selectedGenes, loadingGenes).length > 0) {
      setVegaData(null);
      return;
    }

    const data = populateHeatmapData(cellSets, heatmapSettings, expressionData, selectedGenes, true);
    setDataDebounce(data);
  }, [loadingGenes,
    hidden,
    selectedTracks,
    groupedTracks,
    maxCells,
    properties,
    hierarchy,
    expressionValue]);

  useEffect(() => {
    setMaxCells(Math.floor(width * 0.8));
  }, [width]);

  const handleMouseOver = (...args) => {
    if (args.length < 2) {
      return;
    }

    if ('x' in args[1] && 'y' in args[1]) {
      hoverCoordinates.current = {
        x: args[1].x,
        y: args[1].y,
      };
    }

    if (!args[1].datum) {
      return;
    }

    const { cellId: cellName } = args[1].datum;

    dispatch(
      updateCellInfo(
        {
          cellName,
        },
      ),
    );
  };

  const signalListeners = {
    mouseOver: handleMouseOver,
  };

  if (!selectedGenes || selectedGenes.length === 0) {
    return (
      <Empty
        description={(
          <Text>Add some genes to this heatmap to get started.</Text>
        )}
      />
    );
  }

  if (cellSetsLoading || expressionData.loading.length) {
    return (
      <center>
        <Loader experimentId={experimentId} />
      </center>
    );
  }

  if (!vegaData) {
    return (
      <center style={{ marginTop: height / 2 }}>
        <Skeleton.Image />
      </center>
    );
  }

  if (error || viewError) {
    return (
      <PlatformError
        error={error}
        onClick={() => {
          dispatch(loadGeneExpression(experimentId, selectedGenes, COMPONENT_TYPE));
        }}
      />
    );
  }
  return (
    <div>
      <VegaHeatmap
        spec={vegaSpec}
        data={vegaData}
        showAxes={selectedGenes?.length <= 30}
        rowsNumber={selectedGenes.length}
        signalListeners={signalListeners}
        width={width}
        height={height}
      />
    </div>
  );
};

HeatmapPlot.defaultProps = {
};

HeatmapPlot.propTypes = {
  experimentId: PropTypes.string.isRequired,
  width: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
};

export default HeatmapPlot;

export { HeatmapPlot, COMPONENT_TYPE };
