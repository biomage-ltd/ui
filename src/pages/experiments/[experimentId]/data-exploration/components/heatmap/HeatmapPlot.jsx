import React, { useRef, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { useSelector, useDispatch } from 'react-redux';
import {
  Empty, Spin,
} from 'antd';
import _ from 'lodash';
import spec from '../../../../../../utils/heatmapSpec';
import VegaHeatmap from './VegaHeatmap';
import HeatmapCrossHairs from './HeatmapCrossHairs';
import CellInfo from '../CellInfo';
import PlatformError from '../../../../../../components/PlatformError';
import { updateCellInfo } from '../../../../../../redux/actions/cellInfo';
import { loadGeneExpression } from '../../../../../../redux/actions/genes';

const HeatmapPlot = (props) => {
  const {
    experimentId, width, height,
  } = props;

  const componentType = 'Heatmap';

  const dispatch = useDispatch();

  const loadingGenes = useSelector((state) => state.genes.expression.loading);
  const selectedGenes = useSelector((state) => state.genes.expression.views[componentType]?.data);
  const [vegaData, setVegaData] = useState(null);

  const expressionData = useSelector((state) => state.genes.expression);
  const hoverCoordinates = useRef({});
  const cellSetData = useSelector((state) => state.cellSets);
  const { expressionError: error } = expressionData;
  const viewError = useSelector((state) => state.genes.expression.views[componentType]?.error);

  useEffect(() => {
    if (!selectedGenes || selectedGenes.length === 0) {
      return;
    }

    if (_.intersection(selectedGenes, loadingGenes).length > 0) {
      setVegaData(null);
      return;
    }

    const data = createVegaData(selectedGenes, expressionData);
    setVegaData(data);
  }, [loadingGenes]);

  const createVegaData = (selected, expression) => {
    const data = { cellOrder: [], geneOrder: [], heatmapData: [] };

    data.geneOrder = selected;

    data.cellOrder = [];

    const louvainClusters = cellSetData.hierarchy.filter((clusters) => clusters.key === 'louvain');
    if (louvainClusters.length > 0) {
      const clusterKeys = louvainClusters[0].children;

      clusterKeys.forEach(({ key }) => {
        data.cellOrder.push(...cellSetData.properties[key].cellIds);
      });
    }

    selected.forEach((gene) => {
      if (!expression.data[gene]) {
        return;
      }

      data.heatmapData.push({
        gene,
        expression: expression.data[gene].expression,
      });
    });

    return data;
  };

  const handleMouseOver = (...args) => {
    if (args.length < 2) {
      return;
    }
    if (args[1].datum) {
      const { cellId: cellName, expression, gene: geneName } = args[1].datum;
      dispatch(updateCellInfo({
        cellName, expression, geneName, componentType,
      }));
    }
    if ('x' in args[1] && 'y' in args[1]) {
      hoverCoordinates.current = {
        x: args[1].x,
        y: args[1].y,
      };
    }
  };

  const signalListeners = {
    mouseOver: handleMouseOver,
  };

  if (!selectedGenes || selectedGenes.length === 0) {
    return (
      <center>
        <Empty
          description={(
            <span>
              Please add gene(s) to the heatmap from the Gene list tool
            </span>
          )}
        />
        <HeatmapCrossHairs />
      </center>
    );
  }

  const isHeatmapLoading = () => {
    if (_.intersection(selectedGenes, loadingGenes).length > 0) {
      return true;
    }

    if (!vegaData) {
      return true;
    }

    return false;
  };

  if (isHeatmapLoading()) {
    return (
      <center style={{ marginTop: height / 2 }}>
        <Spin size='large' />
        <HeatmapCrossHairs />
      </center>
    );
  }

  if (error || viewError) {
    return (
      <PlatformError
        description={error}
        onClick={() => {
          if (!isHeatmapLoading()) {
            dispatch(loadGeneExpression(experimentId, selectedGenes, componentType));
          }
        }}
      />
    );
  }

  return (
    <div>
      <VegaHeatmap
        spec={spec}
        data={vegaData}
        showAxes={selectedGenes?.length <= 30}
        rowsNumber={selectedGenes.length}
        defaultWidth={width + 35}
        signalListeners={signalListeners}
        width={width}
        height={height}
      />
      <div className='cell-info-container'>
        <CellInfo
          coordinates={hoverCoordinates}
          componentType={componentType}
        />
      </div>
      <HeatmapCrossHairs />
    </div>
  );
};

HeatmapPlot.defaultProps = {
};

HeatmapPlot.propTypes = {
  experimentId: PropTypes.string.isRequired,
  width: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
};

export default HeatmapPlot;
