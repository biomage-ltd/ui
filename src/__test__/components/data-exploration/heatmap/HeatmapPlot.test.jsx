import React from 'react';
import { configure, mount } from 'enzyme';
import { Provider } from 'react-redux';
import Adapter from 'enzyme-adapter-react-16';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { Empty } from 'antd';
// eslint-disable-next-line import/no-named-as-default
import HeatmapPlot from '../../../../components/data-exploration/heatmap/HeatmapPlot';
import VegaHeatmap from '../../../../components/data-exploration/heatmap/VegaHeatmap';

import { CELL_SETS_LOADING } from '../../../../redux/actionTypes/cellSets';

jest.mock('localforage');
jest.mock('../../../../components/data-exploration/heatmap/VegaHeatmap');
VegaHeatmap.mockImplementation(() => <div>Mocked Vega Heatmap</div>);

const mockStore = configureMockStore([thunk]);
configure({ adapter: new Adapter() });
let component;
const componentType = 'interactiveHeatmap';

const initialState = {
  genes: {
    expression: {
      loading: [],
      data: {
        REALGENE: {
          min: 0,
          max: 1.6,
          expression: [0, 0.4, 0.5, 1.6],
        },
        REALGENE2: {
          min: 0,
          max: 1.6,
          expression: [0, 0.4, 0.5, 1.6],
        },
        REALGENE3: {
          min: 0,
          max: 1.6,
          expression: [0, 0.4, 0.5, 1.6],
        },
      },
      views: {
        [componentType]: {
          data: ['REALGENE'],
          fetching: false,
          error: false,
        },
      },
    },
  },
  cellSets: {
    hierarchy: [
      {
        key: 'louvain',
        children: [
          {
            key: 'louvain-0',
          },
          {
            key: 'louvain-1',
          },
        ],
      },
    ],
    properties: {
      louvain: {
        name: 'louvain clusters',
      },
      'louvain-0': {
        name: 'louvain 0',
        cellIds: new Set([5, 6, 7]),
      },
      'louvain-1': {
        name: 'louvain 1',
        cellIds: new Set([1, 2, 3]),
      },
    },
    hidden: new Set([5]),
  },
  cellInfo: {},
  componentConfig: {
    interactiveHeatmap: {
      config: {
        groupedTracks: ['sample'],
        selectedTracks: ['louvain'],
      },
    },
  },
};

describe('HeatmapPlot', () => {
  afterEach(() => {
    component.unmount();
  });

  it('renders Empty component when no selected gene', () => {
    const store = mockStore({
      ...initialState,
      genes: {
        ...initialState.genes,
        expression: {
          ...initialState.genes.expression,
          views: {
            ...initialState.genes.expression.views,
            [componentType]: {
              ...initialState.genes.expression.views[componentType],
              fetching: false,
              error: false,
              data: [],
            },
          },
          loading: [],
        },
      },
    });

    component = mount(
      <Provider store={store}>
        <HeatmapPlot experimentId='123' width={200} height={200} />
      </Provider>,
    );
    expect(component.find('HeatmapPlot').length).toEqual(1);
    expect(component.find(Empty).length).toEqual(1);
  });

  it('renders Spinner when no expression data', () => {
    const store = mockStore({
      ...initialState,
      genes: {
        ...initialState.genes,
        expression: {
          ...initialState.genes.expression,
          views: {
            ...initialState.genes.expression.views,
            [componentType]: {
              ...initialState.genes.expression.views[componentType],
              fetching: true,
              error: false,
              data: ['REALGENE'],
            },
          },
          loading: ['REALGENE'],
        },
      },
    });

    component = mount(
      <Provider store={store}>
        <HeatmapPlot experimentId='123' width={200} height={200} />
      </Provider>,
    );
    expect(component.find('HeatmapPlot').length).toEqual(1);
    expect(component.find('Loader').length).toEqual(2);
  });

  it('renders Vega heatmap when genes loaded', () => {
    const store = mockStore(initialState);
    component = mount(
      <Provider store={store}>
        <HeatmapPlot experimentId='123' width={200} height={200} />
      </Provider>,
    );
    expect(component.find('HeatmapPlot').length).toEqual(1);
    expect(component.find('VegaHeatmap').length).toEqual(1);
  });

  it('renders error state when expression data errors out', () => {
    const store = mockStore({
      ...initialState,
      genes: {
        ...initialState.genes,
        expression: {
          ...initialState.genes.expression,
          error: 'wow, error!',
        },
      },
    });

    component = mount(
      <Provider store={store}>
        <HeatmapPlot experimentId='123' width={200} height={200} />
      </Provider>,
    );

    expect(component.find('HeatmapPlot').length).toEqual(1);
    expect(component.find(Empty).length).toEqual(1);
  });

  it('renders error state when the view errors out', () => {
    const store = mockStore({
      ...initialState,
      genes: {
        ...initialState.genes,
        expression: {
          ...initialState.genes.expression,
          views: {
            ...initialState.genes.expression.views,
            [componentType]: {
              ...initialState.genes.expression.views[componentType],
              error: true,
            },
          },
        },
      },
    });

    component = mount(
      <Provider store={store}>
        <HeatmapPlot experimentId='123' width={200} height={200} />
      </Provider>,
    );

    expect(component.find('HeatmapPlot').length).toEqual(1);
    expect(component.find(Empty).length).toEqual(1);
  });

  it('dispatches loadCellSets action when no cell sets are in the store', () => {
    const store = mockStore({
      ...initialState,
      cellSets: {
        ...initialState.cellSets,
        hierarchy: [],
        properties: [],
        loading: false,
        error: true,
      },
    });

    component = mount(
      <Provider store={store}>
        <HeatmapPlot experimentId='123' width={200} height={200} />
      </Provider>,
    );

    expect(component.find('HeatmapPlot').length).toEqual(1);
    expect(store.getActions().length).toEqual(1);
    const [loadAction] = store.getActions();

    expect(loadAction.type).toBe(CELL_SETS_LOADING);
    expect(loadAction).toMatchSnapshot();
  });
});
