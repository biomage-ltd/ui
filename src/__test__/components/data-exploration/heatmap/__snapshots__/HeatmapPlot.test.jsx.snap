// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeatmapPlot dispatches loadCellSets action when no cell sets are in the store 1`] = `
Object {
  "type": "cellSets/loading",
}
`;
